lib/instrumentation/undici.js

```diff
@@ -0,0 +1,226 @@
+    segment.start()
+    shim.setActiveSegment(segment)
+    const transaction = segment.transaction
+    const outboundHeaders = Object.create(null)
+    if (agent.config.encoding_key && transaction.syntheticsHeader) {
+      outboundHeaders[NEWRELIC_SYNTHETICS_HEADER] = transaction.syntheticsHeader
+    }
+
+    // TODO: abstract header logic shared with TransactionShim#insertCATRequestHeaders
+    if (agent.config.distributed_tracing.enabled) {
+      if (request.headers && request.headers[SHIM_SYMBOLS.DISABLE_DT]) {
+        logger.trace('Distributed tracing disabled by instrumentation.')
+      } else {
+        transaction.insertDistributedTraceHeaders(outboundHeaders)
+      }
+    } else if (agent.config.cross_application_tracer.enabled) {
+      if (agent.config.encoding_key) {
+        _addCATHeaders(agent, transaction, outboundHeaders)
+      } else {
+        logger.trace('No encoding key found, not adding CAT headers')
+      }
+    } else {
+      logger.trace('CAT disabled, not adding headers!')
+    }
+
+    // eslint-disable-next-line guard-for-in
+    for (const key in outboundHeaders) {
+      request.addHeader(key, outboundHeaders[key])
+    }
+
+    // TODO: Need https://github.com/nodejs/undici/issues/863
+    segment.addAttribute('url', `${request.origin}${url.pathname}`)
+
+    url.searchParams.forEach((value, key) => {
+      segment.addSpanAttribute(`request.parameters.${key}`, value)
+    })
+    segment.addAttribute('procedure', request.method || 'GET')
+    request[SEGMENT] = segment
+    request[PARENT_SEGMENT] = parent
+  })
+
+  diagnosticsChannel.channel('undici:request:headers').subscribe(({ request, response }) => {
+    const activeSegment = request[SEGMENT]
+    if (activeSegment) {
+      activeSegment.addSpanAttribute('http.statusCode', response.statusCode)
+      activeSegment.addSpanAttribute('http.statusText', response.statusText)
+      // If CAT is enabled, grab those headers!
+      if (
+        agent.config.cross_application_tracer.enabled &&
+        !agent.config.distributed_tracing.enabled
+      ) {
+        pullCatHeaders(
+          agent.config,
+          activeSegment,
+          request.origin,
+          response.headers['x-newrelic-app-data']
+        )
+      }
+    }
+  })
+
+  diagnosticsChannel.channel('undici:request:trailers').subscribe(({ request }) => {
+    const activeSegment = request[SEGMENT]
+    const parentSegment = request[PARENT_SEGMENT]
+    if (activeSegment) {
+      activeSegment.end()
+      if (parentSegment) {
+        shim.setActiveSegment(parentSegment)
+      }
+    }
+  })
+
+  diagnosticsChannel.channel('undici:request:error').subscribe(({ request, error }) => {
+    const activeSegment = request[SEGMENT]
+    const parentSegment = request[PARENT_SEGMENT]
+    if (activeSegment) {
+      activeSegment.end()
+      // TODO: this is not IncomingMessage so has no listener count
+      handleError(
+        activeSegment,
+        {
```

Comment:
can we get the actual http req(IncomingMessage)

Reply:
I guess u need it for this?
https://github.com/newrelic/node-newrelic/blob/main/lib/instrumentation/core/http-outbound.js#L172-L175
```
  if (req.listenerCount('error') > 0) {
    logger.trace(error, 'Not capturing outbound error because user has already handled it.')
    return false
  }
```
 AFAIK think there is no way in undici to know if user is handling the error, because it returns the error as the parameter on callback / by rejecting the promise. But i'm not the author of this lib so I don't actually know.
 
 Btw I think its `ClientRequest` here, not `IncominMessage`.
IMO even for the native `http` if there are listeners on the event, does not mean it is actually handles, most of the http libs (e.g axios, got) will set the listener to reject the promise

example:
https://github.com/axios/axios/blob/f3ca6371caa738ba5308d413433d9f676f2e0138/lib/adapters/http.js#L285-L288

So maybe better to set the value to "1" here, not 0